src/callbacks.py

```diff
@@ -73,3 +83,172 @@ def get_prediction_masks(self):
+                for (name, loss_function_one, weight), target in zip(self.loss_function, targets_var):
+                    loss_sum = loss_function_one(outputs_batch, target) * weight
+                outputs.setdefault(self.output_names[0], []).append(outputs_batch.data.cpu().numpy())
+            else:
+                batch_losses = []
+                for (name, loss_function_one, weight), output, target in zip(self.loss_function, outputs_batch, targets_var):
+                    loss = loss_function_one(output, target) * weight
+                    batch_losses.append(loss)
+                    partial_batch_losses.setdefault(name, []).append(loss)
+                    output_ = output.data.cpu().numpy()
+                    outputs.setdefault(name, []).append(output_)
+                loss_sum = sum(batch_losses)
+            partial_batch_losses.append(loss_sum)
+            if batch_id == steps:
+                break
+        self.model.train()
+        average_losses = sum(partial_batch_losses) / steps
+        outputs = {'{}_prediction'.format(name): np.vstack(outputs_) for name, outputs_ in outputs.items()}
+        for name, prediction in outputs.items():
+            outputs[name] = softmax(prediction, axis=1)
+
+        return outputs, average_losses
+
+    def _generate_prediction(self, outputs):
+        data = {'callback_input': {'meta': self.meta_valid,
+                                   'meta_valid': None,
+                                   'target_sizes': self.meta_valid[SIZE_COLUMNS].values,
+                                   },
+                'unet_output': {**outputs}
+                }
+        with TemporaryDirectory() as cache_dirpath:
+            pipeline = self.validation_pipeline(cache_dirpath)
+            output = pipeline.transform(data)
+        y_pred = output['y_pred']
+
+        return y_pred
+
+
+class ModelCheckpointSegmentation(ModelCheckpoint):
+    def __init__(self, metric_name='sum', *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.metric_name = metric_name
+
+    def on_epoch_end(self, *args, **kwargs):
+        if self.epoch_every and ((self.epoch_id % self.epoch_every) == 0):
+            self.model.eval()
+            val_loss = self.get_validation_loss()
+            loss_sum = val_loss[self.metric_name]
+            loss_sum = loss_sum.data.cpu().numpy()[0]
+
+            self.model.train()
+
+            if self.best_score is None:
+                self.best_score = loss_sum
+
+            if (self.minimize and loss_sum < self.best_score) or (not self.minimize and loss_sum > self.best_score) or (
+                    self.epoch_id == 0):
+                self.best_score = loss_sum
+                save_model(self.model, self.filepath)
+                logger.info('epoch {0} model saved to {1}'.format(self.epoch_id, self.filepath))
+
+        self.epoch_id += 1
+
+
+def postprocessing__pipeline_simplified(cache_dirpath):
```

Comment:
@taraspiotr It seems that this pipeline should be different for sigmoid mask/background model and softmax mask/border/background model. Is it so?

Reply:
@jakubczakon That could be true in the future, but at the moment both unet_masks (mask, background) and unet_borders (mask, borders, background) are using softmax.